### ISSUE TITLE
When creating a new Customer Group, it has no access to all categories

### ISSUE THREAD

**Issue by alabaeye** [PrestaShop/PrestaShop#27674]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

When I wanted to create a group on prestashop, I realized that when the group was created, it was automatically added to the carriers and not to the product categories that give access to the sites.

When you create a group it will be automatically added in the "Delivery => Carrier" section, it will be checked in all carriers, however it will not be automatically added in the "Catalogue => Categories" section, which was a problem for me, given that not all new groups have access to all categories but had access to free shipping, which is normally only available to one group.

### Expected behavior

At the creation of a group I wanted the group to be automatically added to the categories of products so that it is access to all the products and not to all the carriers, because they have a specific group for the delivery costs.

### Steps to reproduce

To reproduce the problem you just have to create a group on prestashop and check in your carriers as well as product categories.

However I have solved the problem, I will explain to you how I proceeded to solve it.

For the addition to all product categories, I retrieve in the prestashop "classes" folder, the "Category.php" class, which I duplicate entirely in my "override" folder at the root of my project, and then I resumed the "setNewGroupForHome($idGroup) function", which I duplicate just below in the file, which I called "setNewGroupForAll($idGroup)" with this code:

```
public static function setNewGroupForAll($idGroup)
{
    if (!(int) $idGroup) {
        return false;
    }
    
    $resAll = Db::getInstance()->executeS('SELECT * FROM `' . _DB_PREFIX_ . 'category" cp');
    
    foreach($resAll as $res) {
        Db::getInstance()->execute('INSERT INTO `' . _DB_PREFIX_ . 'category_group' (`id_category`, `id_group`) VALUES (' . (int) $res['id_category'] . ', ' . (int) $idGroup . ')');
    }
    
    return true;
}
```

This code allows me to retrieve all existing product categories and add the group to the group access.

Then I duplicated in the same way the "Group.php" class entirely, in this file I modified the function "add($autodate = true, $null_values = false)", I show you below, the initial code and my modification,

Initial code:

```
public function add($autodate = true, $null_values = false)
{
        Configuration::updateGlobalValue('PS_GROUP_FEATURE_ACTIVE', '1');
        if (parent::add($autodate, $null_values)) {
            Category::setNewGroupForHome((int) $this->id);
            Carrier::assignGroupToAllCarriers((int) $this->id);

            return true;
        }

        return false;
}
```

Code changed: 

```
public function add($autodate = true, $null_values = false)
{
        Configuration::updateGlobalValue('PS_GROUP_FEATURE_ACTIVE', '1');
        if (parent::add($autodate, $null_values)) {
            // I added the line below and put in comment the old ones
            Category::setNewGroupForAll((int) $this->id);
            //Category::setNewGroupForHome((int) $this->id);
            //Carrier::assignGroupToAllCarriers((int) $this->id);

            return true;
        }

        return false;
}
```

### PrestaShop version(s) where the bug happened

1.7.6.9

### PHP version(s) where the bug happened

 7.2.24

**Comment by Hlavtox** [comment]
@alabaeye Hey, I also faced this, even made a script that will check that all categories have all groups checked. :-)

Do you think you could open a PR for this? However, I would add two new options to PS configuration - "Add group to all categories when creating it" or something like this.

**Comment by alabaeye** [comment]
> 

Hello @Hlavtox,

Yes I think I can open a PR about it, I can write the same as in this ticket, adding your idea, to add Prestashop configuration?

**Comment by florine2623** [comment]
Hello @alabaeye ,

You issue might be related to : https://github.com/PrestaShop/PrestaShop/issues/26921

I have reproduced your issue.
Here are the steps to reproduce the issue :
- create a new Customer Group
- go to FO, see that menu is not displayed + go to a category URL (mysite.com/3-clothes)
- see error ⛔ 
<img width="1152" alt="Screenshot 2022-02-15 at 14 36 41" src="https://user-images.githubusercontent.com/16019289/154073013-ca90129b-f186-49c4-8809-809c838dd872.png">
- go to the said category, it is not accessible to my newly created group unless I check it manually ⛔ 
<img width="816" alt="Screenshot 2022-02-15 at 14 36 25" src="https://user-images.githubusercontent.com/16019289/154073192-888deacc-7ddf-42d7-8d7f-55e80322a0d6.png">


You can create a PR here : https://github.com/PrestaShop/PrestaShop/pulls
More info here : https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/pull-requests/
You can explain briefly the issue and how to test it. Do not forget to link this issue to your future PR :)

Thanks!

**Comment by leemyongpakvn** [comment]
I think the current behavior is good enough. Who know what you intend for the new group? Imagine you have a shop with some dangerous tools, and you need a new group for Foe or Children.

**Comment by sengarashish** [comment]
Do we have any update on this bug. Even after manually fixing parent category, child categories are not automatically fixed and they need to be manually updated which is a lot of work.

**Comment by Hlavtox** [comment]
@sengarashish It's not difficult to do, but we should decide if we want to do it automatically, or add a configuration for this.

Ping @PrestaShop/product-council

**Comment by kpodemski** [comment]
@Hlavtox I think this is something what @jf-viguier implemented some time ago, but I might be wrong

**Comment by Hlavtox** [comment]
@kpodemski I think it was only for creating categories. It didn't check all groups when creating a category

**Comment by RosaBenouamer** [comment]
I have the impression that it depends on the merchants' needs. It might be a good idea to set parameters for selecting or not selecting the group in the old categories and carriers. @MatShir your mind ?

**Comment by MatShir** [comment]
Configuration settings are often overlooked, but they can have significant consequences, such as the shared @leemyongpakvn. I think it depends largely on the merchant. One possible solution could be to ask users if they want to associate a new user group with all existing categories and/or carriers when the group is created. This could resolve most use cases. WDYT? (A bit like when we delete a category, we ask how the user wants to proceed in a bulk action command.)

**Comment by Hlavtox** [comment]
Yeah, good idea. We can just use a modal or an input in the form when creating a group and ask.

"Do you want to automatically assign this group to all categories, carriers and other entities?"

**Comment by sengarashish** [comment]
That is a very good idea and I am in favor of it. But if we can also add a
provision to assign the groups to categories during edit with child
categories, it would solve problems for many existing shops.

Regards,
Ashish

**Comment by sitte** [comment]
This fix works for me:

```
INSERT INTO ps_category_group (id_category, id_group)
SELECT c.id_category, g.id_group
FROM ps_category AS c
CROSS JOIN ps_group AS g
LEFT JOIN ps_category_group AS cg
  ON cg.id_category = c.id_category AND cg.id_group = g.id_group
WHERE cg.id_category IS NULL;
```

It adds all id_category and id_group combinations that does not allready exists.

**Comment by boo-code** [comment]
Still reproduces on 9.2.x. Created a group through `AddCustomerGroupCommand`, the same path the back office ends up on, in a shop with 9 categories and 4 carriers:

```
rows in category_group for the new group : 1  (of 9)
rows in carrier_group  for the new group : 4  (of 4)
rows in module_group   for the new group : 0
```

The asymmetry is two adjacent lines in `Group::add()`:

```php
Category::setNewGroupForHome((int) $this->id);      // home category only
Carrier::assignGroupToAllCarriers((int) $this->id); // every carrier
```

So the report is exact: all carriers, one category. The single row is the home category, not the root.

**The product question this ticket was waiting on already has an answer in this thread.** @MatShir proposed asking the merchant at creation time rather than deciding for them, @Hlavtox agreed on the shape ("Do you want to automatically assign this group to all categories, carriers and other entities?"), and that also answers @leemyongpakvn's objection, since nothing would be assigned silently. I do not think anything is blocked on a decision any more.

**What blocks it now is that the form is not migrated.** The customer groups page is only half migrated: the grid is Symfony, but both form actions bounce straight back to the legacy controller -

```php
// CustomerGroupsController::createAction()
return $this->redirect($legacyContext->getAdminLink('AdminGroups', true, ['addgroup' => '']));
// editAction() does the same with updategroup
```

and the `customer_group` feature flag is off by default. `AddCustomerGroupCommand` exists and is exercised by the API layer and by Behat, but nothing builds it from a form. So adding the agreed checkbox today means adding a field to the legacy `AdminGroupsController` HelperForm and its `postProcess` - extending the very code the migration is meant to replace, and it would have to be written twice.

That makes this a natural thing to land **with** the group form migration rather than before it: the command already exists, so the option is a flag on `AddCustomerGroupCommand` plus one checkbox in the new form, and the handler assigns the categories after `create()`. I am happy to prepare that once the form is migrated, or now against the legacy form if maintainers would rather not wait - it is your call which side of the migration it belongs on.

One caution about the workaround above, since people will copy it: @sitte's query cross-joins every group with every category, so it also grants access to groups that were deliberately restricted - which is exactly the case @leemyongpakvn described. If you run it, restrict it to the group you just created:

```sql
INSERT INTO ps_category_group (id_category, id_group)
SELECT c.id_category, <id_group> FROM ps_category c
LEFT JOIN ps_category_group cg ON cg.id_category = c.id_category AND cg.id_group = <id_group>
WHERE cg.id_category IS NULL;
```
